test/layout-test.ts

```diff
@@ -119,6 +251,64 @@ describe('layout', () => {
       assert.deepEqual(breakpoints, [0]);
     });
 
+    it('generates narrow frog prince layout from p. 81 of Digital Typography', () => {
+      const items = frogPrinceItems();
+      // width given on p. 78 of Digital Typography
+      // subtract 1em (18 machine units) from the first line
+      const lineLengths = [372, ...Array(items.length - 1).fill(390)];
+      const breakpoints = breakLines(items, lineLengths);
+      const lines = lineStrings(items, breakpoints);
+      assert.deepEqual(lines, [
+        'In olden times when wishing still helped one,',
+        'there lived a king whose daughters were all beau-',
+        'tiful; and the youngest was so beautiful that the',
+        'sun itself, which has seen so much, was aston-',
+        'ished whenever it shone in her face. Close by the',
+        'king’s castle lay a great dark forest, and under an',
+        'old limetree in the forest was a well, and when',
+        'the day was very warm, the king’s child went out',
+        'into the forest and sat down by the side of the',
+        'cool fountain; and when she was bored she took a',
+        'golden ball, and threw it up on high and caught',
+        'it; and this ball was her favorite plaything. -',
+      ]);
+      const adjRatios = adjustmentRatios(items, lineLengths, breakpoints).map((num) =>
+        Number(num.toFixed(3)),
+      );
+      assert.deepEqual(
+        adjRatios,
+        [0.857, 0.0, 0.28, 1.0, 0.067, -0.278, 0.536, -0.167, 0.7, -0.176, 0.357, 0.049],
```

Comment:
I'm not certain it is guaranteed that these float literals will be converted to the exact same float values as the result of parsing the truncated number, in all environments. If a problem is encountered, `assert.closeTo` could be used.

Reply:
I thought about using `closeTo`, but there wasn't a `deepCloseTo` like there is a `deepEqual` so I would have had to iterate over the list. Happy to switch to that approach if you think it is preferable.

Reply:
I'm happy to leave this as-is. My concern is only theoretical at the moment.